apps/api/src/report/report.service.ts

```diff
@@ -347,81 +380,129 @@ export class ReportService {
    * Get average time spent on the stages, NNAS, BCCNM & NCAS, Recruitment by Health Authorities, and Immigration
    * including overall time from the start to immigration completion and average time to hire.
    *
-   * NOTE: Counting only hired applicants completed immigration
+   * - Each stage counts applicants who completed the stage.
+   *
+   * NNAS
+   * BCCNM & NCAS
+   * Employer
+   *    First Nations Health Authority
+   *    Fraser Health Authority
+   *    Interior Health Authority
+   *    Northern Health Authority
+   *    Providence Health Care Society
+   *    Provincial Health Services Authority
+   *    Vancouver Coastal Health Authority
+   *    Vancouver Island Health Authority
+   * Immigration
+   * Overall
+   * Average time to hire
    *
-   * @param t end date of report, YYYY-MM-DD
+   * @param end date of report, YYYY-MM-DD
+   * @param verbose if true, include sample values and count
    */
-  async getAverageTimeWithEachStakeholderGroup(statuses: Record<string, string>, t: string) {
-    const { to } = this.captureFromTo('', t);
-    this.logger.log(`getAverageTimeWithEachStakeholderGroup: apply filter till ${to} date)`);
-
-    // to get durations of NNAS, BCCNM & NCAS, Immigration, Overall, and Average time to hire
-    const durations = await getRepository(MilestoneDurationEntity)
-      .createQueryBuilder()
-      .where(`hired_at <= :to AND immigrated_at  <= :to`, { to })
-      .getMany();
-
-    // excludes applicants with a negative duration caused by nonlinear milestones
-    const linearDurations = durations.filter(d => !Object.values(d).some(v => v < 0));
-
-    // to get duration of recruitment by Health Authorities
-    const durationByHAs = await getManager()
-      .createQueryBuilder(IENHaPcn, 'ha')
-      .select('ha.title', 'HA')
-      .addSelect(`CAST(COALESCE(ROUND(AVG(md.recruitment), 2), 0) AS FLOAT)`, 'Mean')
-      .addSelect(`COALESCE(MODE() WITHIN GROUP (ORDER BY md.recruitment), 0)`, 'Mode')
-      .addSelect(
-        `COALESCE(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY md.recruitment), 0)`,
-        'Median',
-      )
-      .leftJoin(
-        qb => {
-          return qb
-            .from(MilestoneDurationEntity, 'd')
-            .where(`d.hired_at <= :to AND d.immigrated_at <= :to`, { to })
-            .andWhere(`LEAST(d.pre_screen, interview, reference_check, hired) > 0`);
-        },
-        'md',
-        'ha.title = md.ha',
+  async getReport9(end: string, verbose = false): Promise<object[]> {
+    const { to } = this.captureFromTo('', end);
+    this.logger.log(
+      `Report 9 - Average time with each health authority: apply filter till ${to} date)`,
+      'REPORT',
+    );
+    const table = await this.reportUtilService.getMilestoneTable(to);
+
+    // set health authorities who hired applicants
+    const idHaMap = await this.reportUtilService.getHiredApplicantHAs();
+    Object.entries(idHaMap).forEach(([id, ha]) => (table[id].ha = ha));
```

Comment:
Trying to understand this forEach and the following for loop.  How are these doing anything different besides checking if its truthy? Unless I am misunderstanding something, could this be done all inside the for loop below? 
Something like,  `for(...) { table[id].ha = ha; if(table[id]) table[id].ha = ha}`